core/arch/arm/include/sm/optee_smc.h

```diff
@@ -93,6 +93,12 @@
 #define OPTEE_SMC_SHM_CACHED		1
 
 /*
+ * Privileged call flag. Used by supplicant. Asks OP-TEE to use reserved
+ * thread if all regular threads are used.
+ */
+#define OPTEE_SMC_FLAGS_PRIVILEGED	BIT(0)
```

Comment:
One extra privileged thread isn't enough, one extra privileged thread per normal thread is actually needed to avoid deadlocks. That also lead me into the idea that if during RPC registering of SHM is needed it can be done using the suspended thread currently doing the RPC. If we only allow registration of SHM in this case I don't think we'll have a problem with stack overruns.

However, it will result in some extra ping-pong and it may be a bit complicated to find out which thread did the RPC and do a call inside that.

I'm afraid I've forgotten how `OPTEE_MSG_ATTR_TYPE_NEXT_NESTED` worked, but I think the best solution would be to return of `OPTEE_MSG_RPC_CMD_SHM_ALLOC` to carry the information needed to register the shared memory. Do you really needed `OPTEE_MSG_ATTR_TYPE_NEXT_NESTED` for that?